src/tir/transforms/unify_thread_binding.cc

```diff
@@ -0,0 +1,126 @@
+    if (attr->attr_key != attr::thread_extent && attr->attr_key != attr::virtual_thread) {
+      return StmtMutator::VisitStmt_(attr);
+    }
+
+    // Step 1. Fetch the old IterVar.
+    IterVar old_iter_var = Downcast<IterVar>(attr->node);
+    IterVar new_iter_var;
+
+    // Step 2. See if an IterVar for this kind of thread binding was created before. If so, we use
+    // the created IterVar. Otherwise, we create a new IterVar for this thread binding and store the
+    // IterVar in mapping `thread_tag2iter_var_map_`.
+    Map<String, IterVar>::iterator it = thread_tag2iter_var_map_.find(old_iter_var->thread_tag);
+    if (it != thread_tag2iter_var_map_.end()) {
+      new_iter_var = (*it).second;
+      CHECK(ExprDeepEqual()(old_iter_var->dom->extent, (*it).second->dom->extent))
+          << "ValueError: All loops that are bound to `" << old_iter_var->thread_tag
+          << "` should have the same extent. However, there are two loops with extent "
+          << (*it).second->dom->extent << " and " << old_iter_var->dom->extent
+          << ", which are not equal";
+    } else {
+      ObjectPtr<IterVarNode> p_new_iter_var = make_object<IterVarNode>(*old_iter_var.get());
+      p_new_iter_var->var = Var(old_iter_var->thread_tag);
+      new_iter_var = IterVar(p_new_iter_var);
+      // We don't unify thread bindings of "vthread".
+      if (old_iter_var->thread_tag != "vthread") {
+        thread_tag2iter_var_map_.Set(old_iter_var->thread_tag, new_iter_var);
+      }
+    }
+
+    // Step 3. We will substitute the occurrences of the old variable in the old IterVar with the
+    // new variable in further mutation. Thus, we store the mapping entry.
+    var_substitution_map_.Set(old_iter_var->var, new_iter_var->var);
+
+    // Step 4. Mutate recursively, and update the AttrStmt with the new IterVar.
+    AttrStmt new_attr = Downcast<AttrStmt>(StmtMutator::VisitStmt_(attr));
+    ObjectPtr<AttrStmtNode> p_new_attr = CopyOnWrite(new_attr.get());
+    p_new_attr->node = new_iter_var;
+    return Stmt(p_new_attr);
+  }
+
+  PrimExpr VisitExpr_(const VarNode* var) final {
+    // If this variable appears as a key in `var_substitution_map_`, we substitute it with its
+    // corresponding value in the mapping.
+    Map<Var, Var>::iterator it = var_substitution_map_.find(GetRef<Var>(var));
+    return it != var_substitution_map_.end() ? (*it).second : GetRef<Var>(var);
+  }
+
+  /*!
+   * \brief A mapping from a thread tag to its corresponding IterVar that is shared by all
+   * occurrences of the thread tag
+   * */
+  Map<String, IterVar> thread_tag2iter_var_map_;
+  /*! \brief A mapping from old variables to new variables, which is used for substitution */
+  Map<Var, Var> var_substitution_map_;
+};
+
+PrimFunc UnifyThreadBinding(PrimFunc f) {
+  // Only apply this pass to TIR that is not from TE schedules
+  if (!IsFromLegacyTESchedule(f)) {
+    PrimFuncNode* fptr = f.CopyOnWrite();
+    fptr->body = ThreadBindingUnifier::Unify(f);
```

Comment:
Let's also tweak the signature of `ThreadBindingUnifier::Unify` here

```suggestion
    fptr->body = ThreadBindingUnifier::Unify(std::move(f->body));
```

Reply:
No problem. Also curious about the reason 🙂.